File: PhaseJumpUnity/Assets/phasejumppro/UI/System/UISystem+Select.cs
using System;
using System.Collections.Generic;

/*
RATING: 5 stars
Has Unit Tests
CODE REVIEW: 6/10/23
*/
namespace PJ
{
    /// <summary>
    /// Manages selection state for objects
    /// </summary>
    public partial class UISystem
    {
        protected HashSet<SelectHandler> selection = new();

        public HashSet<SelectHandler> Selection
        {
            get => selection;
            set
            {
                HashSet<SelectHandler> removedSelectHandlers = new(selection);
                removedSelectHandlers.ExceptWith(value);

                foreach (var selectHandler in removedSelectHandlers)
                {
                    selectHandler.IsSelected = false;
                }

                foreach (var selectHandler in value)
                {
                    selectHandler.IsSelected = true;
                }

                this.selection = value;
            }
        }

        public void UpdateSelectionFor(SelectHandler selectHandler)
        {
            if (selectHandler.IsSelected)
            {
                selection.Add(selectHandler);
            }
            else
            {
                selection.Remove(selectHandler);
            }
        }
    }
}

File: PhaseJump/PhaseJump/Classes/UI/System/UISystem+Select.cpp
#include "SelectHandler.h"
#include "UIWorldSystem.h"

using namespace std;
using namespace PJ;

void UIWorldSystem::SetSelection(OrderedSet<SP<SelectHandler>> const& value) {
    auto removedSelectHandlers = selection - value;

    for (auto& selectHandler : removedSelectHandlers) {
        selectHandler->SetIsSelected(false);
    }

    for (auto& selectHandler : value) {
        selectHandler->SetIsSelected(true);
    }

    selection = value;
}

void UIWorldSystem::UpdateSelectionFor(SP<SelectHandler> selectHandler) {
    if (selectHandler->IsSelected()) {
        selection.insert(selectHandler);
    } else {
        selection.erase(selectHandler);
    }
}
